build/ublue-os-just/50-akmods.just

```diff
@@ -28,3 +28,121 @@ configure-broadcom-wl ACTION="prompt":
+          esac
+
+      done
+    else
+        luk=$(echo "${luks[0]}" | sed -e 's/\\n//g')
+        [[ ! "$luk"  =~ [:alnum:] ]] && echo "No devices with crypto active ... exiting" && exit
+    fi
+    [[ $device = */dev/* ]] || device=/dev/$device
+    clear;
+    echo "Your device is $device  with volume $luk"
+    read -p "confirm choices y/n  or quit q " choice
+    clear
+    case $choice in
+      [Qq]*) exit;;
+      [Yy]*) : ;;
+      *) echo "invalid choice, exiting"; exit;;
+    esac
+
+    read -p "do you need to enrol a new key?  if yes insert key before answering  y/n " choice
+    case $choice in
+        [Yy]* ) systemd-cryptenroll --fido2-device=auto --fido2-with-user-verification=true $device;;
+        *) echo "not enrolling a new key" ;;
+    esac
+    ### edit the crypttab like for the luks volume if needed
+    luksline=$(grep -r $luks /etc/crypttab)
+    if [[ ! "$luksline" = *"fido2-device"* ]] ; then
+        opts=$(echo $luksline | awk '{print $4}')
+        newopts=$(echo $luksline | sed "s/$opts/fido2-device=auto,$opts/g")
+        sed -i "s/$luksline/$newopts/g" /etc/crypttab
+        echo "updated crypttab with fido"
+    fi
+    dracutfile="/etc/dracut.conf.d/fido2.conf"
+    if [ ! -f $dracutfile ]; then
+        echo 'add_dracutmodules+=" fido2 "
+    force_drivers+=" fido2 "' > $dracutfile
+        echo "created fido2 dracut.conf"
+    else
+        mods="$(grep -r "add_dracutmodules+=" $dracutfile)"
+        if [[ $mods = *"dracutmodules+="* ]] ; then
+            if [[ ! $mods = *"fido2"* ]] ; then
+                sed -i 's/add_dracutmodules+="/add_dracutmodules+=" fido2 /g' "$dracutfile"
+            fi
+        else
+            echo 'add_dracutmodules+=" fido2 "' >> $dracutfile
+        fi
+        drivers="$(grep -r "force_drivers+=" $dracutfile)"
+        if [[ $drivers = *"force_drivers+="* ]] ; then
+            if [[ ! $drivers = *"fido2"* ]] ; then
+                sed -i 's/force_drivers+="/force_drivers+=" fido2 /g' "$dracutfile"
+            fi
+        else
+            echo 'force_drivers+=" fido2 "' >> $dracutfile
+        fi
+        echo "updated fido2 dracut.conf"
+    fi
+    echo "
+
+    Set the slot priority of the of the password and fido
+    0: fido preferred, password ignored  | WARNING !!  This will ONLY allow the fido to unlock
+    1: fido preferred, password normal   | This will allow the password to be used if fido isnt present
+    "
+    read -p ": " choice
+    case $choice in
+        [0] ) cryptsetup config --key-slot 0 --priority ignore $device ; cryptsetup config --key-slot 1 --priority prefer $device;;
+        [1] ) cryptsetup config --key-slot 0 --priority normal $device ; cryptsetup config --key-slot 1 --priority prefer $device;;
+        *) echo "invalid option,  leaving slot priorities the same" ;;
+    esac
+    echo "
+
+    "
+    read -p "Enable the initramfs to rebuild ?  y/n " choice
+    case $choice in
+        [Yy]* ) rpm-ostree initramfs --enable;;
```

Comment:
unsure if this is needed if dracut modules are already enabled, but will have to double check if main images have dracut modules enabled by default. (will be able to check that tomorrow)

if they are not enabled or **if you need to** issue a rebuild then what you will have to do (thanks to how `rpm-ostree initramfs` functions) get the current initramfs arguments (with quotes in the right places for each `--arg=`) from `rpm-ostree -b | grep Initramfs | sed 's/Initramfs: //'` and rebuild the previously used `rpm-ostree initramfs` command to avoid wiping out the current args.

Reply:
it seems that running this at this point makes it rebuild anyway, but if you have a better method im happy to change it. from my testing (on a bazzite build) a rebuild was needed for this module to work as intended.

and it seems that, according to rpm-ostree,    --arg  appends the line to the dracut arguments   and --enable Enable regenerating initramfs locally using dracut

it shouldnt be wiping anything out when doing a rebuild afaik  each deployment seems to retain the initramfs args from the last

Reply:
Only Bazzite enables initramfs by default.

Additionally unsure if the force_drivers is necessary for fido. The dracutmodule should suffice for pulling it in. At least it does that for tpm2.

Should we have these security devices just specified in a file in /usr/lib/dracut.conf.d. if they are already present, we would only have to enable initramfs generation, do the key enrollment, and specify the fido device as a karg.

Key enrollment would be the only item that needs elevated privileges.

Reply:
> it seems that running this at this point makes it rebuild anyway, but if you have a better method im happy to change it. from my testing (on a bazzite build) a rebuild was needed for this module to work as intended.
> 
> and it seems that, according to rpm-ostree, --arg appends the line to the dracut arguments and --enable Enable regenerating initramfs locally using dracut
> 
> it shouldnt be wiping anything out when doing a rebuild afaik each deployment seems to retain the initramfs args from the last

from experience if you happen to lets say have `--arg="-I /etc/crypttab /etc/modprobe.d/nvidia.conf"` as the initramfs arguments and you run `rpm-ostree initramfs --enable`, `rpm-ostree initramfs` in its infinite wisdom will remove `--arg="-I /etc/crypttab /etc/modprobe.d/nvidia.conf"` as it was never supplied and the nvidia.conf modprobe never gets added and the nvidia driver does not get loaded. we had to work around this problem in bazzite which is why we have our own rebuilder script that gets triggered by `sudo touch /etc/bazzite/initramfs/rebuild` however it is currently not compatible with other images.

Reply:
> Only Bazzite enables initramfs by default.
> 
> Additionally unsure if the force_drivers is necessary for fido. The dracutmodule should suffice for pulling it in. At least it does that for tpm2.
> 
> Should we have these security devices just specified in a file in /usr/lib/dracut.conf.d. if they are already present, we would only have to enable initramfs generation, do the key enrollment, and specify the fido device as a karg.
> 
> Key enrollment would be the only item that needs elevated privileges.

i like the sound of this idea, i should also just look into the idea of a `initramfs-rebuild` that issues a rebuild, nothing else specifically for situations like this. However ideally everything would use initramfs-etc but that would require us to get the other bits sorted out first.